Allow using callee-save registers when WASM SIMD is enabled

justinmichaud opened the pull request:
#### 63b6c42ff2b8191437f4257dd5ca251d91851d80
<pre>
Allow using callee-save registers when WASM SIMD is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=250151">https://bugs.webkit.org/show_bug.cgi?id=250151</a>
rdar://103929372

Reviewed by Michael Saboff.

When a function allocates a vector tmp in an FP calle-save register (ARM64 only),
it must treat the top half of that register as volatile. To work around
a bug, we currently avoid using those registers at all.

1) Fix a bug caused by an incorrect rebase that claimed that we did not clobber
fp callee-saves at all when SIMD is used in a function.

2) Avoid using linear scan with SIMD. It is not optimal and barely tested.

3) Ensure that patchpoints correctly claim that they only clobber the top
of callee-save fp registers.

4) Fix some minor assertion failures and noisy test output.

5) Add a microbenchmark to prove that SIMD still regresses OSR

6) Through manual inspection, verify that we can now generate the following code:

Patch &amp;Patchpoint0, %x0, %q8 b@16

instead of

MoveFloat %q8, -44(%fp), b@11
Patch &amp;Patchpoint0, %x0, -44(%fp), b@16

iff %q8 is not used as a vector.

* JSTests/wasm/v8/graceful_shutdown.js:
(CompileFunctionsTest):
* JSTests/wasm/v8/import-function.js:
(testCallPrint.fakePrint):
(testCallPrint):
* JSTests/wasm/v8/many-modules.js:
* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackByLinearScan.cpp:
(JSC::B3::Air::allocateRegistersAndStackByLinearScan):
* Source/JavaScriptCore/b3/air/AirCode.cpp:
(JSC::B3::Air::Code::Code):
* Source/JavaScriptCore/b3/air/AirGenerate.cpp:
(JSC::B3::Air::prepareForGeneration):
* Source/JavaScriptCore/b3/air/AirLowerAfterRegAlloc.cpp:
(JSC::B3::Air::lowerAfterRegAlloc):
* Source/JavaScriptCore/b3/air/AirTmpWidth.cpp:
(JSC::B3::Air::TmpWidth::recompute):
* Source/JavaScriptCore/jit/RegisterSet.h:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::emitCatchImpl):
(JSC::Wasm::AirIRGenerator64::addThrow):
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType&gt;::finalizeEntrypoints):
(JSC::Wasm::ExpressionType&gt;::addLoop):
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::emitCatchPrologueShared):

Canonical link: <a href="https://commits.webkit.org/258757@main">https://commits.webkit.org/258757@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ccd3061659c3a8ad2dc5eafae698adb4b77ef5ec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102845 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11963 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/35871 "Build was cancelled. Recent messages:Cleaned up git repository; Running apply-patch; Checked out pull request; Compiled WebKit (cancelled)") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/112099 "Build was cancelled. Recent messages:Cleaned up git repository; Running apply-patch; Checked out pull request; Updated wpe dependencies; compiling") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172319 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12980 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2872 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95093 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109769 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108619 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/35871 "Build was cancelled. Recent messages:Cleaned up git repository; Running apply-patch; Checked out pull request; Compiled WebKit (cancelled)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37611 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/35871 "Build was cancelled. Recent messages:Cleaned up git repository; Running apply-patch; Checked out pull request; Compiled WebKit (cancelled)") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/93052 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5416 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/35871 "Build was cancelled. Recent messages:Cleaned up git repository; Running apply-patch; Checked out pull request; Compiled WebKit (cancelled)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/89413 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/3102 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5564 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2569 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29676 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11583 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/35871 "Build was cancelled. Recent messages:Cleaned up git repository; Running apply-patch; Checked out pull request; Compiled WebKit (cancelled)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/92336 "Built successfully") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/6008 "Build was cancelled. Recent messages:Cleaned up git repository; Checked out pull request; Verified commit is squashed; Reviewed by Michael Saboff; Validated commit message; Compiled WebKit (cancelled)") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7309 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20679 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3203 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

msaboff commented on JSTests/wasm/v8/graceful_shutdown.js:
```diff
@@ -25,6 +25,6 @@ load("wasm-module-builder.js");
   WebAssembly.compile(buffer).then(
-    () => { print("success")},
-    () => { print("failed"); });
+    () => { /*print("success")*/},
```
Can we eliminate the comment or at least the `print()` call?

msaboff commented on JSTests/wasm/v8/many-modules.js:
```diff
@@ -37,7 +37,7 @@ const modules = [];
 module_bytes.set(Array(sentinel.length).fill(_ => kExprNop), sentinel_position);
 for (let i = 0; i < num_modules; ++i) {
-  if (i % 50 == 0) print(i);
+  // if (i % 50 == 0) print(i);
```
Please eliminate this line.

msaboff reviewed (approved):
r=me with minor nits

justinmichaud committed:
Allow using callee-save registers when WASM SIMD is enabled
https://bugs.webkit.org/show_bug.cgi?id=250151
rdar://103929372

Reviewed by Michael Saboff.

When a function allocates a vector tmp in an FP calle-save register (ARM64 only),
it must treat the top half of that register as volatile. To work around
a bug, we currently avoid using those registers at all.

1) Fix a bug caused by an incorrect rebase that claimed that we did not clobber
fp callee-saves at all when SIMD is used in a function.

2) Avoid using linear scan with SIMD. It is not optimal and barely tested.

3) Ensure that patchpoints correctly claim that they only clobber the top
of callee-save fp registers.

4) Fix some minor assertion failures and noisy test output.

5) Add a microbenchmark to prove that SIMD still regresses OSR

6) Through manual inspection, verify that we can now generate the following code:

Patch &Patchpoint0, %x0, %q8 b@16

instead of

MoveFloat %q8, -44(%fp), b@11
Patch &Patchpoint0, %x0, -44(%fp), b@16

iff %q8 is not used as a vector.

* JSTests/wasm/v8/graceful_shutdown.js:
(CompileFunctionsTest):
* JSTests/wasm/v8/import-function.js:
(testCallPrint.fakePrint):
(testCallPrint):
* JSTests/wasm/v8/many-modules.js:
* Source/JavaScriptCore/b3/air/AirAllocateRegistersAndStackByLinearScan.cpp:
(JSC::B3::Air::allocateRegistersAndStackByLinearScan):
* Source/JavaScriptCore/b3/air/AirCode.cpp:
(JSC::B3::Air::Code::Code):
* Source/JavaScriptCore/b3/air/AirGenerate.cpp:
(JSC::B3::Air::prepareForGeneration):
* Source/JavaScriptCore/b3/air/AirLowerAfterRegAlloc.cpp:
(JSC::B3::Air::lowerAfterRegAlloc):
* Source/JavaScriptCore/b3/air/AirTmpWidth.cpp:
(JSC::B3::Air::TmpWidth::recompute):
* Source/JavaScriptCore/jit/RegisterSet.h:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::emitCatchImpl):
(JSC::Wasm::AirIRGenerator64::addThrow):
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType>::finalizeEntrypoints):
(JSC::Wasm::ExpressionType>::addLoop):
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::emitCatchPrologueShared):

Canonical link: https://commits.webkit.org/258757@main
